### PR TITLE
Add UDPRoute type for UDP support

### PR DESCRIPTION
**Pull request by JeremyOT**
Adds basic UDPRoute and UDP protocol support

#227

### CODE REVIEW

**Comment by hbagdi** [comment]
+1

Two points to clarify:
1. As per previous discussions in our office hours, UDPRoute should fall under custom or extended conformance. Is this still the case? If yes, then we should mention this somewhere in docs, preferrably not inside Godoc comments.
2. We have laid out compatibility between listeners [here](https://github.com/kubernetes-sigs/service-apis/blob/70efb5f7b9ee4a9d4192584f8e1f3115e7349902/apis/v1alpha1/gateway_types.go#L69-L73). The current language indicates that TCP and UDP listeners cannot be compatible with others (which seem correct). I want to confirm if that's indeed correct.

**Comment by jpeach** [comment]
@JeremyOT What are the network semantics of a UDP route? Does it forward the packet unmodified, or does it generate a new  UDP packet whose source address is the intermediary? Is anything implied about the return path?

If we don't want to define the semantics yet, I think that's OK (we can just see what useful things implementations do) too.

A slightly different question too; could this be used to handle custom protocols on top of UDP? It seems like the extensionRef on the match could be used for that purpose?

**Comment by SantoDE** [comment]
The SMI group is currently trying to answer a subpart of that question as well. What information can we take to route UDP?

Currently, in this proposal here, the `UDPRouteMatch` is empty which would mean, that we would need one listener per application to proxy to, right? Or am I wrong?

On the lastest SMI approach (https://github.com/servicemeshinterface/smi-spec/issues/156), we were discussing whether what additional matchers to add or not. Currently, the SMI UDP Spec has ports only as well. However, we discussed about the possibility to match on the SNI as well, at least for UDP connections that utilize TLS. It's probably a bit too advanced for a basic implementation but might be worth keeping it in our heads?

**Comment by JeremyOT** [comment]
If we don't want to define the semantics yet, I think that's OK (we can just see what useful things implementations do) too.

I'd lean toward wait and see. I'm not sure what all we'd want to specify - though those all seem like reasonable candidates.

> A slightly different question too; could this be used to handle custom protocols on top of UDP? It seems like the extensionRef on the match could be used for that purpose?

Yeah, an extensionRef seems like the right way to try things out. If we wanted to make anything more official, we'd probably want another *Route type

**Comment by JeremyOT** [comment]
> Currently, in this proposal here, the `UDPRouteMatch` is empty which would mean, that we would need one listener per application to proxy to, right? Or am I wrong?
> 
> The SMI group is currently trying to answer a subpart of that question as well. What information can we take to route UDP?
> 
> On the lastest SMI approach ([servicemeshinterface/smi-spec#156](https://github.com/servicemeshinterface/smi-spec/issues/156)), we were discussing whether what additional matchers to add or not. Currently, the SMI UDP Spec has ports only as well. However, we discussed about the possibility to match on the SNI as well, at least for UDP connections that utilize TLS. It's probably a bit too advanced for a basic implementation but might be worth keeping it in our heads?

I think it would make sense to follow #238 here. In that case SNI would probably belong on a different kind of route (DTLSRoute?) which should definitely consider, but for now we should probably stick with port and see what evolves. Extensions may help experiment in the mean time?

**Comment by beriberikix** [comment]
I've been tracking the PR closely (and SMI @SantoDE) and are very excited to see progress! I thought it might be helpful to give a few examples. I'm coming the IoT, which has *a lot* custom protocols, some of which we're currently implementing in a cloud native way. Some common characteristics:

- IP-based
- Custom L7 protocol on `UDP`/`TCP`
- URIs & ports for routing
- Explicit protocols for routing too, esp. to distinguish between secure modes (ex. `coap://` vs. `coaps://`)
- App-layer security

Note not every protocol meets these criteria, like `MQTT` relies on `TLS` at the `TCP` layer...some are not even IP-based (but may be compatible with L2.)

So the first use-case for us would be `CoAP` ([RFC7252](https://tools.ietf.org/html/rfc7252).) We'd start with insecure `CoAP`, this `UDPRoute` and something with `extensionRef`. Then when `DTLSRoute` gets defined, we can implement `coaps://` + `extensionRef`. Eventually we would implement a `CoAPRoute` & `CoAPSRoute`.

Is that the right way to think about it?

**Comment by bowei** [comment]
Yes -- we expect that specific routing will expand beyond vanilla (see SNIRoute for example)

**Comment by hbagdi** [comment]
> So the first use-case for us would be CoAP (RFC7252.) We'd start with insecure CoAP, this UDPRoute and something with extensionRef. Then when DTLSRoute gets defined, we can implement coaps:// + extensionRef. Eventually we would implement a CoAPRoute & CoAPSRoute.

Spot on. Something we haven't documented but implementers are free to and even encouraged to define such new XRoute resources. We will pull them into extended or core conformance as convergence progresses across vendors.

File: apis/v1alpha1/udproute_types.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+// UDPRouteStatus defines the observed state of UDPRoute
+type UDPRouteStatus struct {
+	GatewayRefs []GatewayObjectReference `json:"gatewayRefs" protobuf:"bytes,1,rep,name=gatewayRefs"`
+}
+
+// +genclient
+// +kubebuilder:object:root=true
+
+// UDPRoute is the Schema for the udproutes API
```
Comment (jpeach) [inline apis/v1alpha1/udproute_types.go:36]:
```suggestion
UDPRoute is the Schema for the udproutes API.
```

File: apis/v1alpha1/udproute_types.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	// group), or an implementation may support other resources (for
+	// example, resource "myroutetargets" in group "networking.acme.io").
+	// Omitting or specifying the empty string for both the resource and
+	// group indicates that the resource is "services".  If the referent
+	// cannot be found, the "InvalidRoutes" status condition on any Gateway
+	// that includes the UDPRoute will be true.
+	ForwardTo *ForwardToTarget `json:"forwardTo" protobuf:"bytes,1,opt,name=forwardTo"`
+
+	// ExtensionRef is an optional, implementation-specific extension to the
+	// "action" behavior.  The resource may be "configmaps" (use the empty
```
Comment (jpeach) [inline apis/v1alpha1/udproute_types.go:70]: This is the same language used in TCPRoute, so it's fine for this PR. However, as written it's not clear whether the extension *is* an action (i.e. does something) or merely *modifies* an action (i.e. extends ForwardTo in some capacity). Maybe we should clarify.

File: apis/v1alpha1/udproute_types.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// UDPRouteSpec defines the desired state of UDPRoute
+type UDPRouteSpec struct {
+	// Rules are a list of UDP matchers and actions.
+	Rules []UDPRouteRule `json:"rules" protobuf:"bytes,1,rep,name=rules"`
+}
+
+// UDPRouteStatus defines the observed state of UDPRoute
```
Comment (jpeach) [inline apis/v1alpha1/udproute_types.go:28]:
```suggestion
// UDPRouteStatus defines the observed state of UDPRoute.
```

File: apis/v1alpha1/udproute_types.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// UDPRouteSpec defines the desired state of UDPRoute
```
Comment (jpeach) [inline apis/v1alpha1/udproute_types.go:22]:
```suggestion
// UDPRouteSpec defines the desired state of UDPRoute.
```

**Comment by jpeach** [comment]
> 2. We have laid out compatibility between listeners [here](https://github.com/kubernetes-sigs/service-apis/blob/70efb5f7b9ee4a9d4192584f8e1f3115e7349902/apis/v1alpha1/gateway_types.go#L69-L73). The current language indicates that TCP and UDP listeners cannot be compatible with others (which seem correct). I want to confirm if that's indeed correct.

That seems overly restrictive, but we should probably file an issue for it and tackle it separately. Maybe we need to remove the rules around protocol compatibility and let implementations figure it out 🤷

**Comment by robscott** [comment]
@hbagdi here's my take on your questions:
> 1. As per previous discussions in our office hours, UDPRoute should fall under custom or extended conformance. Is this still the case? If yes, then we should mention this somewhere in docs, preferrably not inside Godoc comments.

Agreed. I believe this should probably be under extended conformance but think that we can clarify that in a follow up that defines conformance for each route type.

> 2. We have laid out compatibility between listeners here. The current language indicates that TCP and UDP listeners cannot be compatible with others (which seem correct). I want to confirm if that's indeed correct.

Agree with @jpeach above that this is likely more restrictive than necessary but that we can file an issue to tackle it separately.

**Comment by robscott** [comment]
Thanks for your work on this @JeremyOT! I think all feedback has been addressed here, so this LGTM.

/lgtm

File: apis/v1alpha1/udproute_types.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+	Spec   UDPRouteSpec   `json:"spec,omitempty" protobuf:"bytes,3,opt,name=spec"`
+	Status UDPRouteStatus `json:"status,omitempty" protobuf:"bytes,4,opt,name=status"`
+}
+
+// UDPRouteRule is the configuration for a given rule.
+type UDPRouteRule struct {
+	// Match defines which packets match this rule.
+	//
+	// +optional
```
[the remark this answers is not in the document]
Comment (bowei) [reply apis/v1alpha1/udproute_types.go:49]: +1, let's copy over the language in TCPRoute if that is applicable.

**Comment by robscott** [comment]
Thanks!

/lgtm

**Comment by robscott** [comment]
Thanks!

/lgtm

**Pull request merged**